packages/runtime/container-runtime/src/containerRuntime.ts

```diff
@@ -1466,11 +1478,126 @@ export class ContainerRuntime extends EventEmitter
+            let dataStoreEntryChanged: boolean = false;
+
+            // Get the path of the referenced object relative to the dataStoreId and add it to the
+            // referencedFluidObjects in its references.
+            const dataStoreIdRelativePath = nextRoute.slice(dataStoreId.length + 1);
+            if (dataStoreIdRelativePath !== "") {
+                assert(references.referencedRoutes);
+                references.referencedRoutes.add(dataStoreIdRelativePath);
+                dataStoreEntryChanged = true;
+            }
+
+            // If this data store has not been visited yet, mark it as visited.
+            if (!references.visited) {
+                references.visited = true;
+                dataStoreEntryChanged = true;
+            }
+
+            // If the entry has changed, update it in the dataStoreReferencesMap.
+            if (dataStoreEntryChanged) {
+                dataStoreReferencesMap.set(dataStoreId, references);
+            }
+
+            // If this data store has been already visited, ignore it.
+            if (visited) {
+                continue;
+            }
+
+            // Get the routes referenced by this data store and add them to the referencedRoutesStack.
+            const referencedRoutes = this.getDataStoreReferencedRoutes(dataStoreId);
+            referencedRoutesStack.push(...referencedRoutes);
+        }
+
+        const deletedDataStores: string[] = [];
+        // Now we have a map of all the data stores that are referenced and the routes that are referenced in it.
+        // Iterate through all the data stores in this runtime and:
+        // 1. If they are not referneced, mark them for collection.
+        // 2. If they are referenced, pass the list of routes in them that are referenced so it can collect
+        //    unreferenced objects.
+        this.contexts.forEach((value: FluidDataStoreContext) => {
+            // Ignore the scheduler.
+            if (value.id === taskSchedulerId) {
+                return;
+            }
+
+            // If the data store does exists in the dataStoreReferencesMap, it is no referenced. Mark it for collection.
+            // Else, get the list of the referenced routes in it and ask it to collect the unreferenced objects.
+            if (!dataStoreReferencesMap.has(value.id)) {
+                deletedDataStores.push(value.id);
+                value.setReferenced(false);
+            } else {
+                const references = dataStoreReferencesMap.get(value.id);
+                assert(references);
+                value.setReferenced(true);
+                value.collectUnreferencedFluidObjects([...references.referencedRoutes]);
+            }
+        });
+        return deletedDataStores;
+    }
+
+    private getDataStoreReferencedRoutes(dataStoreId: string): string[] {
+        const dataStore = this.contexts.get(dataStoreId);
+        assert(dataStore);
```

Comment:
This assumes that ALL data stores are loaded, right?
I think that can't be the case, we need pay-per-play solution where data store is loaded only if it was changed, or maybe for legacy files where we do not have info from last summary stored in snapshot about routes used in it.

Reply:
This is something that needs some thought and work. Currently, when you call a data store to get snapshot, its all or nothing - it will either take entire snapshot or not. We need a mechanism that says something like - the only thing that changed is whether its referenced or not, so only update that part of the summary tree.

Reply:
No, it's not all or nothing. Data store returns snapshot, but it can reuse bits and pieces (sub-trees) from previous snapshot. So it's pay-per-play on every level. Yes, it eventually looks like full snapshot, but storage make it a reality by flattening it (i.e. resolving old references).
That said, we should not load data store at all if it did not change - that's the issue I'm raising here.